File: examples/java/example/apps/BitfieldSub.java
package example.apps;

import java.io.*;
import java.util.Arrays;
import zcm.zcm.*;
import javazcm.types.bitfield_t;

public class BitfieldSub implements ZCMSubscriber
{
    ZCM zcm;
    ZCM.Subscription s;

    public BitfieldSub()
    {
        zcm = ZCM.getSingleton();
        zcm.start();
        sub();
    }

    public void sub()
    {
        s = zcm.subscribe("BITFIELD", this);
    }

    public void unsub()
    {
        zcm.unsubscribe(s);
    }

    public void messageReceived(ZCM zcm, String channel, ZCMDataInputStream ins)
    {
        System.out.println("Received message on channel " + channel);

        try {
            bitfield_t msg = new bitfield_t(ins);
            assert (msg.field1 == -1);
            assert (msg.field2[0][0] == -1);
            assert (msg.field2[0][1] ==  0);
            assert (msg.field2[0][2] == -1);
            assert (msg.field2[0][3] ==  0);
            assert (msg.field2[1][0] == -1);
            assert (msg.field2[1][1] ==  0);
            assert (msg.field2[1][2] == -1);
            assert (msg.field2[1][3] ==  0);
            assert (msg.field3 == -1);
            assert (msg.field4 == -3);
            assert (msg.field5 == 7);
            assert (msg.field6 == 1);
            assert (msg.field7 == 3);
            assert (msg.field8_dim1 == 0);
            assert (msg.field8_dim2 == 0);
            assert (msg.field8.length == 0);
            assert (msg.field9 == ~(1 << 27) + 1);
            assert (msg.field10 == (((long)1 << 52) | 1));
            assert (msg.field11 == 3);
            for (int i = 0; i < 2; ++i) {
                for (int j = 0; j < 2; ++j) {
                    for (int k = 0; k < 2; ++k) {
                        for (int l = 0; l < 2; ++l) {
                            assert (msg.field12[i][j][k][l] == k + l + 1);
                        }
                    }
                }
            }
            assert (msg.field15 == -60);
            assert (msg.field16 == 2);
            assert (msg.field18 == 15);
            assert (msg.field19 == 68);
            assert (msg.field20 == 2);
            assert (msg.field22 == msg.FIELD22_TEST);
            assert (msg.field23 == msg.FIELD23_TEST);
            assert (msg.field24 == msg.FIELD24_TEST);
            assert (msg.field25 == 3);
            assert (msg.field26 == (byte)0xff);
            assert (msg.field27 == 3);
            assert (msg.field28 == 0x7f);
            assert (msg.field29 == 3);
            assert (msg.field30 == 0x7fff);
            assert (msg.field31 == 0xf);
            assert (msg.field32 == 0x7fffffff);
            assert (msg.field33 == 0xf);
            assert (msg.field34 == 0x7fffffffffffffffL);
        } catch (IOException ex) {
            System.out.println("Exception: " + ex);
        }
    }

    static void sleep(long ms)
    { try { Thread.sleep(ms); } catch (InterruptedException ex) {} }

    public static void main(String[] args)
    {
        BitfieldSub s = new BitfieldSub();
        while(true) {
            sleep(1000);
        }
    }
}

File: examples/cpp/BitfieldSub.cpp
#include <stdio.h>
#include <unistd.h>
#include <string>
#include <zcm/zcm-cpp.hpp>
#include "types/bitfield_t.hpp"
using std::string;

class Handler
{
    public:
        ~Handler() {}

        void handleMessage(const zcm::ReceiveBuffer* rbuf,
                           const string& chan,
                           const bitfield_t *msg)
        {
            printf("Received message on channel \"%s\":\n", chan.c_str());
            printf("%d\n", msg->field1);
            printf("[");
            for (int j = 0; j < 2; ++j) {
                printf("[");
                for (int i = 0; i < 4; ++i) {
                    printf("%d, ", msg->field2[j][i]);
                }
                printf("], ");
            }
            printf("]\n");
            printf("%d\n", msg->field3);
            printf("%d\n", msg->field4);
            printf("%d\n", msg->field5);
            printf("%d\n", msg->field6);
            printf("%d\n", msg->field7);
            printf("%d\n", msg->field8_dim1);
            printf("%d\n", msg->field8_dim2);
            printf("[");
            for (int j = 0; j < msg->field8_dim1; ++j) {
                printf("[");
                for (int i = 0; i < msg->field8_dim2; ++i) {
                    printf("%d, ", msg->field8[j][i]);
                }
                printf("]");
            }
            printf("]\n");
            printf("%d\n", msg->field9);
            printf("%ld\n", msg->field10);
            printf("%d\n", msg->field11);
            printf("[");
            for (int l = 0; l < 3; ++l) {
                printf("[");
                for (int k = 0; k < 2; ++k) {
                    printf("[");
                    for (int j = 0; j < 2; ++j) {
                        printf("[");
                        for (int i = 0; i < 2; ++i) {
                            printf("%d, ", msg->field12[l][k][j][i]);
                        }
                        printf("], ");
                    }
                    printf("], ");
                }
                printf("], ");
            }
            printf("]\n");
            printf("%d\n", msg->field15);
            printf("%d\n", msg->field16);
            printf("%d\n", msg->field18);
            printf("%d\n", msg->field19);
            printf("%d\n", msg->field20);
            printf("%d\n", msg->field21);
            printf("%d\n", msg->field22);
            printf("%d\n", msg->field23);
            printf("%d\n", msg->field24);
            printf("%d\n", msg->field25);
            printf("%d\n", msg->field26);
            printf("%d\n", msg->field27);
            printf("%d\n", msg->field28);
            printf("%d\n", msg->field29);
            printf("%d\n", msg->field30);
            printf("%d\n", msg->field31);
            printf("%d\n", msg->field32);
            printf("%ld\n", msg->field33);
            printf("%ld\n", msg->field34);
        }
};

int main(int argc, char *argv[])
{
    zcm::ZCM zcm {""};
    if (!zcm.good())
        return 1;

    Handler handlerObject;
    zcm.subscribe("BITFIELD", &Handler::handleMessage, &handlerObject);
    zcm.run();

    return 0;
}
